### PR TITLE
prosemirror-commands: Add types Command and Keymap

### PR DESCRIPTION
**Pull request by mmorearty**
Previously, no types were specified for a Command or for a Keymap. This
meant that (a) in prosemirror-commands, the types for pcBaseKeymap,
macBaseKeymap, and baseKeymap were not very well specified, and (b) in
prosemirror-keymap, there was some duplication of types.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://prosemirror.net/docs/ref/#commands and https://prosemirror.net/docs/ref/#keymap
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by mmorearty** [comment]
@timjb you actually suggested something like this in a [comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44705#discussion_r426287907) on an earlier PR of mine; I didn't do it at that time, but it turns out that now it would be helpful to me!

File: types/prosemirror-commands/index.d.ts
Comment (mmorearty) [inline types/prosemirror-commands/index.d.ts:25]: It's not really clear to me if these types should go here, or in prosemirror-keymap. They are used by both.

File: types/prosemirror-commands/index.d.ts
Comment (timjb) [inline types/prosemirror-commands/index.d.ts:238]: I want to point out that this is a breaking change. After reading the README and some discussions in issues, it seems to me that it's recommended for versions of type definitions to

* always track the library version
* follow semvar

which together imply that breaking changes to the type definitions are only possible when the the corresponding library publishes a new major version. I don't agree with this policy because it means that one must willingly tolerate imprecise or plain wrong type definitions for an indefinite time (e.g. there has been no major release of `prosemirror-commands` since 1.0.0 was released in October 2017). Therefore I'm ignoring this policy and approve this breaking change.

File: types/prosemirror-commands/index.d.ts
Comment (timjb) [reply types/prosemirror-commands/index.d.ts:25]: I think that prosemirror-commands is the best place for the `Command` interface. Unfortunately, moving the `Keymap` interface to prosemirror-keymap would introduce a cycle. Thus, the current solution seems to be the best one to me.

File: types/prosemirror-commands/index.d.ts
Comment (timjb) [inline types/prosemirror-commands/index.d.ts:15]: Are there any reasons for using `any` instead of `unknown` here? If not, could you please replace `any` with `unknown`?

File: types/prosemirror-commands/index.d.ts
Comment (timjb) [inline types/prosemirror-commands/index.d.ts:23]: Are there any reasons for using `any` instead of `unknown` here? If not, could you please replace `any` with `unknown`? (I believe that this would also require using the type `Keymap<any>` for `baseKeymap`, `pcBaseKeymap` and `macBaseKeymap` because otherwise using one of these values would not be possible when something of type `Keymap<S>` for some concrete schema type `S` is required.)

**Review by timjb — CHANGES_REQUESTED** [review]
Some remarks and a question on the choice of `any`/`unknown`.

File: types/prosemirror-commands/index.d.ts
Comment (mmorearty) [reply types/prosemirror-commands/index.d.ts:15]: `unknown` is not compatible with the constraint type `Schema`. If I try `<S extends Schema = unknown>`, the compiler objects, "Type 'unknown' does not satisfy the constraint 'Schema'. (2344)"

I think `<S extends Schema = Schema>` is worth considering. `Schema` itself is defined as `Schema<N extends string = any, M extends string = any>`.

I'm a little hesitant to do that in this PR, though, because throughout the rest of all of the prosemirror packages, there is currently a consistent pattern of using `<S extends Schema = any>`.

I did a little experimentation, and I do quite like `<S extends Schema = Schema>`. I'm not quite sure in practice where it is useful, but it does feel more correct. For example, `EditorState<S extends Schema>` exposes a property `schema: S`. As things are written today, since the declaration says `EditorState<S extends Schema = any>`, my editor does not provide autocomplete in this scenario:

```ts
declare const editorState: EditorState;
editorState.schema.    // <-- I typed "." but I get no hints, since `schema` is of type `any`
```

If I change the declaration to `EditorState<S extend Schema = Schema>`, I do get autocomplete hints:

Screen Shot 2020-07-02 at 9 29 34 AM

I would suggest leaving this line the way I have it in this PR, and then doing a separate PR to change to `<S extends Schema = Schema>` throughout all of the ProseMirror packages.

The same could be done for any other types that use `= any`. For example, `Schema` itself could be changed to `Schema<N extends string = string, M extends string = string>`.

File: types/prosemirror-commands/index.d.ts
Comment (timjb) [reply types/prosemirror-commands/index.d.ts:23]: (see other comment)

File: types/prosemirror-commands/index.d.ts
Comment (timjb) [reply types/prosemirror-commands/index.d.ts:15]: I'm okay with using `s extends Schema = any` for consistency reasons. Let's experiment with `s extends Schema = Schema` in another change!

**Review by timjb — APPROVED** [review]
Thanks for your contribution! :-)

**Comment by mmorearty** [comment]
@RyanCavanaugh does the typescript-bot need a nudge? This PR has been approved.

**Comment by mmorearty** [comment]
Ready to merge

**Comment by mmorearty** [comment]
Argh, how do I get this PR un-stuck? It still has a "Revision needed" label.

**Comment by mmorearty** [comment]
Regarding this PR being stuck, Ryan Cavanaugh [says](https://twitter.com/SeaRyanC/status/1280924907921010688): "Thanks for the ping; I've passed this on to the engineers working on the DT automation this week"

**Comment by elibarzilay** [comment]
(If anyone sees this, please don't merge, so I can catch the bug...)

**Comment by elibarzilay** [comment]
@mmorearty: so the thing is that @timjb left a "changes requested" review, and then an approved review without changes to the PR, so the bot still counted the old one too...  I'll make it ignore an old review for the same author even if it's for the same headref.

**Comment by elibarzilay** [comment]
@mmorearty: one more note -- had the above been fixed, it would have still required a maintainer approval since it's a single PR that edits more than one package.  In the future, it would be quicker to have multiple PRs for different packages.

File: types/prosemirror-commands/index.d.ts
Comment (aried3r) [inline types/prosemirror-commands/index.d.ts:19]: Hey! I was wondering, shouldn't it be `dispatch?` and `view?`?

Most of the command type definitions in this file use:

```TypeScript
state: EditorState<S>,
dispatch?: (tr: Transaction<S>) => void,
view?: EditorView<S>
```

and the docs say:

> A command function takes an editor state, _optionally_ a `dispatch` function that it can use to dispatch a transaction and _optionally_ an `EditorView` instance.

If we add the following code to the `prosemirror-commands-tests.ts` file, this wouldn't compile.

```TypeScript
Object.keys(commands.baseKeymap).forEach(key => {
    if (keymap[key]) {
        keymap[key] = commands.chainCommands(keymap[key], commands.baseKeymap[key]);
    } else {
        keymap[key] = commands.baseKeymap[key];
    }
});
```

(I forgot where this code came from, but apparently it's [a thing](https://github.com/search?q=Object.keys%28baseKeymap%29.forEach&type=Code))

Using the definition from this PR I get:

```
ERROR: 26:46  expect                         TypeScript@next compile error:
Argument of type 'Command<any>' is not assignable to parameter of type '(p1: EditorState<any>, p2?: ((tr: Transaction<any>) => void) | undefined, p3?: EditorView<any> | undefined) => boolean'.
  Types of parameters 'dispatch' and 'p2' are incompatible.
    Type '((tr: Transaction<any>) => void) | undefined' is not assignable to type '(tr: Transaction<any>) => void'.
      Type 'undefined' is not assignable to type '(tr: Transaction<any>) => void'.
```

If I change it to `dispatch?` and `view?` it works.

Am I doing something wrong here? Maybe I'm confusing what is a `Command` and what isn't.

Comment (timjb) [reply types/prosemirror-commands/index.d.ts:19]: @aried3r You are right. It's important that commands can handle the case where `dispatch` and `view` are undefined, because that's used to check whether a command can be applied:

> When no dispatch callback is passed, the command should do a 'dry run', determining whether it is applicable, but not actually doing anything.

We should change that.

Comment (timjb) [reply types/prosemirror-commands/index.d.ts:19]: @aried3r I've made a PR containing a fix: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45972

**Pull request merged**
